[WIP] SRVKS-604: traffic mgmt docs update (#36233)
openshift/openshift-docs: modules/serverless-blue-green-deploy.adoc

[id="serverless-blue-green-deploy_{context}"]
= Routing and managing traffic by using a blue-green deployment strategy

You can safely reroute traffic from a live version of an app to a new version, by using a link:https://en.wikipedia.org/wiki/Blue-green_deployment[blue-green deployment strategy].

.Procedure

. Create and deploy an app as a Knative service.
. Find the name of the first revision that was created when you deployed the service, by viewing the output from the following command:
+
[source,terminal]
----
$ oc get ksvc <service_name> -o=jsonpath='{.status.latestCreatedRevisionName}'
----

> . Add the following YAML to the service `spec` to send inbound traffic to the revision:

Author: I don't think this is actually specified as required in the template, but have added it anyway.

+
[source,yaml]
----
...
spec:
  traffic:
    - revisionName: <first_revision_name>
      percent: 100 # All traffic goes to this revision
...
----
. Verify that you can view your app at the URL output you get from running the following command:
+
[source,terminal]
----
$ oc get ksvc <service_name>
----
. Deploy a second revision of your app by modifying at least one field in the `template` spec of the service and redeploying it. For example, you can modify the `image` of the service, or an `env` environment variable. You can redeploy the service by applying the service YAML file, or by using the `kn service update` command if you have installed the `kn` CLI.
. Find the name of the second, latest revision that was created when you redeployed the service, by running the command:
+
[source,terminal]
----
$ oc get ksvc <service_name> -o=jsonpath='{.status.latestCreatedRevisionName}'
----
+
At this point, both the first and second revisions of the service are deployed and running.
. Update your existing service to create a new, test endpoint for the second revision, while still sending all other traffic to the first revision:
+

> .Example of updated service spec

Author: updated the headings

[source,yaml]
----
...
spec:
  traffic:
    - revisionName: <first_revision_name>
      percent: 100 # All traffic is still being routed to the first revision
    - revisionName: <second_revision_name>
      percent: 0 # 0% of traffic routed to the second revision
      tag: v2 # A named route
...
----
+
After you redeploy this service by reapplying the YAML resource, the second revision of the app is now staged. No traffic is routed to the second revision at the main URL, and Knative creates a new service named `v2` for testing the newly deployed revision.

> . Get the URL of the new service for the second revision, by running the command:

Author: re #1, updated

re #2, they're not really the same
> I guess the earlier step combined retrieving the URL and viewing the URL.

That's correct, but I don't think these are the same, since the first one is just the service in general, the second one is specific to a revision. Since SMEs and QE have already signed off on the technical content as is, I would prefer not to modify this unless there is a good reason to revisit later.

[id="serverless-blue-green-deploy_{context}"]
= Routing and managing traffic by using a blue-green deployment strategy

You can safely reroute traffic from a live version of an app to a new version, by using a link:https://en.wikipedia.org/wiki/Blue-green_deployment[blue-green deployment strategy].

.Procedure

. Create and deploy an app as a Knative service.
. Find the name of the first revision that was created when you deployed the service, by viewing the output from the following command:
+
[source,terminal]
----
$ oc get ksvc <service_name> -o=jsonpath='{.status.latestCreatedRevisionName}'
----
. Add the following YAML to the service `spec` to send inbound traffic to the revision:
+
[source,yaml]
----
...
spec:
  traffic:
    - revisionName: <first_revision_name>
      percent: 100 # All traffic goes to this revision
...
----
. Verify that you can view your app at the URL output you get from running the following command:
+
[source,terminal]
----
$ oc get ksvc <service_name>
----

> . Deploy a second revision of your app by modifying at least one field in the `template` spec of the service. For example, you can modify the `image` of the service, or an `env` environment variable.

Reviewer: I'm thinking if this step and the next one are redundant. I might be wrong but, they look similar and it seems to me just one or another are enough to get revision updated.

[id="serverless-blue-green-deploy_{context}"]
= Routing and managing traffic by using a blue-green deployment strategy

You can safely reroute traffic from a live version of an app to a new version, by using a link:https://en.wikipedia.org/wiki/Blue-green_deployment[blue-green deployment strategy].

.Procedure

. Create and deploy an app as a Knative service.
. Find the name of the first revision that was created when you deployed the service, by viewing the output from the following command:
+
[source,terminal]
----

> $ oc get configurations <service_name> -o=jsonpath='{.status.latestCreatedRevisionName}'

Reviewer 2: ```suggestion
$ oc get ksvc <service_name> -o=jsonpath='{.status.latestCreatedRevisionName}'
```

So we don't introduce `configurations` unnecessarily here.

----
. Add the following YAML to the service `spec` to send inbound traffic to the revision:
+

> .Example Route

Reviewer 2: Should this be `Example Service`?

Author: I think this heading can be removed altogether actually since I reworded this step and made it just a `spec` excerpt. Good catch, thanks

[source,yaml]
----
...
spec:
  traffic:
    - revisionName: <first_revision_name>
      percent: 100 # All traffic goes to this revision
...
----
. Verify that you can view your app at the URL output you get from running the following command:
+
[source,terminal]
----

> $ oc get route <route_name>

Reviewer 2: Stray occurence of route. This can be `oc get service` to stick to that abstraction.

Author: earlier we use `ksvc` and here it would be `service` - can we choose one of these maybe for use with `oc` and stick with it, or is there a reason it'd be different and not `ksvc` here too?

Reviewer 2: `oc get service` won't work, because the K8s Service will be picked then.

Reviewer 2: opps, sorry. Yeah I was wrong above :joy: it should be ksvc.

Author: Thanks 🙂 I'll update them

----
. Deploy a second revision of your app by modifying at least one field in the `template` spec of the service. For example, you can modify the `image` of the service, or an `env` environment variable.
. Redeploy the service by applying the updated `Service` resource. You can do this by applying the service YAML file, or by using the `kn service update` command if you have installed the `kn` CLI.
. Find the name of the second, latest revision that was created when you redeployed the service, by running the command:
+
[source,terminal]
----

> $ oc get configurations <service_name> -o=jsonpath='{.status.latestCreatedRevisionName}'

Reviewer 2: ```suggestion
$ oc get ksvc <service_name> -o=jsonpath='{.status.latestCreatedRevisionName}'
```

same as above

----
+
At this point, both the first and second revisions of the service are deployed and running.

> . Update your existing route to create a new, test endpoint for the second revision, while still sending all other traffic to the first revision:

Reviewer 2: ```suggestion
. Update your existing service to create a new, test endpoint for the second revision, while still sending all other traffic to the first revision:
```

+
.Example of updated route
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Route
metadata:
  name: <route_name>

>   namespace: default

Reviewer 2: Should we drop this too? Also update the heading to say `updated service`?

Author: updated this a bit to make it consistent with earlier example

spec:
  traffic:
    - revisionName: <first_revision_name>
      percent: 100 # All traffic is still being routed to the first revision
    - revisionName: <second_revision_name>
      percent: 0 # 0% of traffic routed to the second revision
      tag: v2 # A named route
----
+

> After you redeploy this route by reapplying the YAML resource, the second revision of the app is now staged. No traffic is routed to the second revision at the main URL, and Knative creates a new route named `v2` for testing the newly deployed revision.

Reviewer 2: ```suggestion
After you redeploy this service by reapplying the YAML resource, the second revision of the app is now staged. No traffic is routed to the second revision at the main URL, and Knative creates a new route named `v2` for testing the newly deployed revision.
```

> . Get the URL of the new route for the second revision, by running the command:
> +
> [source,terminal]
> ----
> $ oc get route <route_name> --output jsonpath="{.status.traffic[*].url}"
> ----

Reviewer 2: ```suggestion
. Get the URL of the new route for the second revision, by running the command:
+
[source,terminal]
----
$ oc get service <service_name> --output jsonpath="{.status.traffic[*].url}"
----
```

+
You can use this URL to validate that the new version of the app is behaving as expected before you route any traffic to it.

> . Update your existing route resource again, so that 50% of traffic is being sent to the first revision, and 50% is being sent to the second revision:
> +
> .Example of updated route
> [source,yaml]
> ----
> apiVersion: serving.knative.dev/v1
> kind: Route
> metadata:
>   name: <route_name>
>   namespace: default

Reviewer 2: ```suggestion
. Update your existing service resource again, so that 50% of traffic is being sent to the first revision, and 50% is being sent to the second revision:
+
.Example of updated service
[source,yaml]
----
```

spec:
  traffic:
    - revisionName: <first_revision_name>
      percent: 50
    - revisionName: <second_revision_name>
      percent: 50
      tag: v2
----

> . Once you are ready to route all traffic to the new version of the app, update the route again to send 100% of traffic to the second revision:
> +
> .Example of updated Route
> [source,yaml]
> ----
> apiVersion: serving.knative.dev/v1
> kind: Route
> metadata:
>   name: <route_name>
>   namespace: default
> spec:
>   traffic:
>     - revisionName: <first_revision_name>
>       percent: 0
>     - revisionName: <second_revision_name>
>       percent: 100
>       tag: v2
> ----

Reviewer 2: ```suggestion
. Once you are ready to route all traffic to the new version of the app, update the service again to send 100% of traffic to the second revision:
+
.Example of updated Service
[source,yaml]
----
spec:
  traffic:
    - revisionName: <first_revision_name>
      percent: 0
    - revisionName: <second_revision_name>
      percent: 100
      tag: v2
----
```

[id="serverless-blue-green-deploy_{context}"]
= Routing and managing traffic by using a blue-green deployment strategy

You can safely reroute traffic from a live version of an app to a new version, by using a link:https://en.wikipedia.org/wiki/Blue-green_deployment[blue-green deployment strategy].

.Procedure

. Create and deploy an app as a Knative service.
. Find the name of the first revision that was created when you deployed the service, by viewing the output from the following command:
+
[source,terminal]
----
$ oc get configurations <service_name> -o=jsonpath='{.status.latestCreatedRevisionName}'
----
. Define a `Route` object to send inbound traffic to the revision:
+

> .Example Route

Reviewer 2: Why are we using `Route` instead of `Service` here? I think we've been using Service everywhere else in the docs thus far?

Author: It came from the docs upstream: https://knative.dev/development/developer/serving/traffic-management/#routing-and-managing-traffic-with-bluegreen-deployment

If it needs to change we should update it in both places; can you open an issue for upstream docs if we change it here?
Also, how would "defining a service" send traffic to a revision? Wouldn't it _need_ to be a route, or would it be setting the same spec just on a service instead?

Reviewer 2: Yes the same spec works inside the `traffic` section of the Service and that imo should be the default way of doing things (as everything we document downstream is based around `Service`)

Author: Updated to mention just adding the `traffic` section to the `spec` of the service.
